# 变量
- 程序=数据结构+算法
- 变量就是可以重复使用 的一个亮
- 变量命名规则
    - 可以包含数字，大小写字母，下划线或者更多，但是我们不推荐除了前三种内容之外的符号
    - 数字不可以打头
        - 4man， 5 for 时不可以的
    - 一般在python中， 以下划线开头的内容有特殊含义， 不建议使用
        - 比如 _age_,理论上可以，强烈不推荐，包括但不限于一个下划线开头
    - 大小可不一致
    - 推荐：
         - 使用具有固定含义的英文或者缩写，srv=server skt=socket,一般以posix命名规则
         - 驼峰命名法
             - 名称以单词自动连接，且每个单词首字母大写
                 - 大驼峰，每个单词第一个字母均大写（类）
                     - 在oython里给类命名用此法
                 - 小驼峰，单词第一个字母小写，其余大写
                 - posix写法（普通变量）
                    - 多个单词全部下划线连接，都用小写
    - 保留字和关键字
        - class ， def ，break， for
        - 变量名必须避开这些词

In [1]:
# 查看关键字的方法
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


## 变量声明
- var_name = var_value # =表示赋值
- var1 = var2 = var3 = var_value  
- 从右往左看
- var1，var2, var3 = v1, v2, v3

In [9]:
# 定义变量age, 把18放入变量age中
age = 18
print(age)
age1 = age2 = age3 = 18 
print(age1)
print(age2)
print(age3)

age4, age5, age6 = 12, 21, 45
print(age4)
print(age5)
print(age6)

18
18
18
18
12
21
45


## 变量类型
- 严格意义上讲，只有一个类型
- 标准上讲，数据类型有六种
    - 数字类型Number
    - 字符串类型str
    - 列表list
    - 元祖tuple
    - 字典 dictionary
    - 集合 set

## Number数字类型
- 数字类型没有大小限制
### 1. 整数
- 没有小数部分，表示个数的数字
- 自然数，0，负自然数
- 整数分进制
    - 二进制
        - 机器蝉蛹
        - 是计算机唯一理解的数字
        - 表示为0b开头的0，1代码
        - 例如0b01101, 0b10001
        
    - 八进制
        - 不常用
        - 以0o开头的包含0-7的数字
    - 十六进制
        - 逢16进1
        - 包含0-9， a-f
        - 数字上以0x开头表示
    - 十进制
        - 18， 89， 178

In [15]:
age = 18
print(18)

age = 0b10010
print(age)

age = 0o22
print(age)

age = 0x12
print(age)

18
18
18
18


### 浮点数
- 小数
- 计算机里边表示浮点数可以有条件省略
- 3.1415
- 3.
- 0.5
- .5
- 科学计数法
    - 定义根数学定义一致
    - 小写使用e/E后面跟证书表示10的指数
    - 176.34 = 1.7634e2
    - 0.876 = 8.76e-1

In [18]:
height = 156.34
print(height)

shili = 0.2
print(shili)
shili = .6
print(shili)

156.34
0.2
0.6


### 复数complex
- 定义同数学一致
- 一个由实部和须部构成的数字
- 虚部用j/J表示
- 5+3j
- 3j
-  （3j）


### 布尔值
- 表示真假的值
- 只有俩个值， True/False
- 布尔值可以跟数字直接操作

In [24]:
# 布尔值操作
age = 18 + True
print(age)

19


## 字符串
- 用来表示一串文字信息
- 如何表示
    - 单引号
    - 双引号
    - 三引号（三三单，三三双）
        - 能够扩起多行信息
        - 汉语状态下符号不对

In [25]:
 love = "我爱王小静"
print(love)

# 单双引号交错使用的例子
# <a style="color:red"> 哈哈哈</a>
# 用引号表示上面字符串
html = 'a style="color:red">哈哈哈</a>'
# 单引号双引号只能表示一行
    

我爱王小静


In [26]:
html = """
asdasdsdasdsadasdasd
"""
print(html)


asdasdsdasdsadasdasd



# 字符串
- 转移字符
- 格式化
- 内建函数

## 转移字符
- 用一个特殊的方法表示出一系列不方便写出的内容，比如回车键，换行键，退格键
- 借助反斜杠字符，一旦字符串中出现反斜杠，则反斜杠后面一个或几个字符表示已经不是原来的意思了，进行了转义
- 在字符串中，一旦出现反斜杠字符就要加倍小心，可能有转义副出现
- 不同操作系统下对操作有不同的表示
    - windows：\n
    - os, linux: \r\n

In [1]:
s = "i love lixiaojing"
print(s)

i love lixiaojing


In [5]:
# 如果想表示出Let's go ,
# 1, 可以用嵌套引号，即外层用双引号
# 2，转移字符
s = "Let's go"
print(s)
ss = 'Let\'s go'
print(ss)
# \\ = \
sss = "c:\\user"
print(sss)
# 回车换行符 \r\n
s1 = "I love \r\n wangxiaojing"
print(s1)

Let's go
Let's go
c:\user
I love 
 wangxiaojing


## 字符串的格式化
- 把字符串按照一定个事进行打印或者填充
- 格式化有俩种方法
    - 利用百分号（%）
    - 利用format函数

In [6]:
s = "xxx 您好， 我们图灵学院开课了，免费的Python 呀，不学白不学，学了也白学"

### 利用百分号格式化
- 在字符串中，利用%表示一个特殊的含义，表示对字符串进行格式化
- %d: 此处应该放入一个整数
- %s: 表示此处应放入一个字符串

In [7]:
s = "I love %s"
# 下面打印，直接把%s作为字符串的一个内容打印出来了
print(s)

I love %s


In [8]:
# 要用格式化就再加一个%并加入要替换内容
print("I love %s"%"wangxiaojing")
print( s%"wangxiaojing")

I love wangxiaojing
I love wangxiaojing


In [9]:
s = "I am %d years old"
print(s)
print(s%18)


I am %d years old
I am 18 years old


In [12]:
s = "I am %s, i am %d years old"
print(s)
# 留下几个站坑的，就必须跟上几个，要么全有药么没有
print(s%("tuilingxueyuan",18))

I am %s, i am %d years old
I am tuilingxueyuan, i am 18 years old


### format函数进行格式化字符串
- 推荐使用这种方法
- 再使用上，以{}和：代替%号，后面用format带参数完成

In [15]:
s = "I love {}".format("lixiaojing")
print(s)
# 按序号进行替换顺序
s = "Yes , i am {1} years old , I love {0} and i am {1} years old".format("Tulingueyuan", 18)
print(s)

I love lixiaojing
Yes , i am 18 years old , I love Tulingueyuan and i am 18 years old


## None
- 表示什么都没有
- 如果函数没有返回值，可以返回None
- 用来占位置
- 用来解除变量的绑定

# 表达式
- 又一个或者几个数字或者变量和运算符组成的一行代码
- 通常会返回一个结果
# 运算符
- 由一个以上的值经过变化得到新值的过程叫运算
- 用于运算的符号叫运算符
- 运算符分类：
    - 算数运算符
    - 比较或者关系运算符
    - 赋值运算符
    - 位运算
    - 成员运算
    - 身份运算符
    - 逻辑运算符
    

## 算数运算符
- 进行运算的符号
- python 没有自增减运算符

In [3]:
# +，-，*，/
a = 9+3-2*2/6
print(a)
# python 2 和 3 中，除号(/)结果可能不一致，此处以3为准
# % 取余运算
# 俩个数字相处应该有商数有余数
# % 只会得到玉树
a = 9 % 4
print(a)
# // 表示取商运算，也叫地板除
a = 9//4
print(a)

# ** 表示幂运算
a = 9 ** 4
print(a)

11.333333333333334
1
2
6561


## 比较运算符
- 对俩个变量或者值进行比较的运算符
- 比较的结果是布尔值，即True/False

In [4]:
# == , 等于号
a = 3 ** 4
# 下面语句执行顺序
# 1， 计算a == 80
# 2, 把结果放入b 中
b = a == 80
print(b)


# ！= 不等于
print(9 != 8)

# > , <, >=, <=

False
True


## 赋值运算符


In [5]:
# = ， 赋值
a = 0 
c = a = 4
# +=, 是缩写。
a = 0 
a += 7 # a = a + 7 的缩写
print(a)

# 所有数学运算符都有缩写形式
# -=， *=， /=， //=， %=， **=，都是缩写形式

7


# 逻辑运算符
- 对布尔值进行计算的符号
- and 逻辑与
- or 逻辑或
- not 逻辑非
- python中逻辑运算符没有异或运算
- 运算规则：
    - and 看作乘法， or看作加法
    - True 看作1， False 看作0
    - 则逻辑运算就能转换成数学运算
    - 最后结果如果是0则为false， 否则为True
- 逻辑运算段路问题
    - 逻辑运算式，把斋运算顺序计算，一旦能够确定整个式子未来的值，则不再进行计算，直接返回
    

In [8]:
# 逻辑运算符案例
a = True
b = False
c = True
# 一下式子等价于 d = 1 * 0 + 1
d = a and b or c
print(d)
d = a or b and a
print(d)

True
True


In [9]:
# 逻辑短路的案例
# 下面的逻辑表达式，a 的值一定是True ,则运行的or的时候，整个表达式不再向下计算
a = True or xxxxxxx

# 下面表达式，如果xxx中包含赋值表达式，则结果很难预测
# 代码实例（伪代码）
a = a or (b =9) and 6
# 假定砂锅面表达式没有语法错误
# 则b最终的值应该是0而不是9
print(b)
print(a)





SyntaxError: invalid syntax (<ipython-input-9-ed2ee781c1c9>, line 6)

# 成员运算符号
- 用来检测某一个变量是否是另一个变量的成员
- in
- not in 

In [10]:
# 案例
l = [1, 2, 3, 4, 5]
a = 7
b = a in l
print(b)
a = 4 
print(a in l)

False
True


## 身份运算
- is : 用来检测俩个变量是否是同一个变量
    - 语法就是 var1 is var2
- is not : 不是同一变量
    

In [12]:
a = 9 
b = 9
print(a is b)

a = "I love wangxiaojing"
b = "I love wangxiaojing"
print(a is b)

True
False


## 运算符的优先级的问题
- 永远记住，括号具有最高优先级
- 优先级表格，略。。


# 程序结构
- 顺序
- 分支
- 循环

# 分支 
- 分支的基本语法

            if 条件表达式：
                语句1
                语句2
                语句3
                ......
- 条件表达式就是计算结果必须为布尔值的表达式
- 表达式后面冒号不能少
- 主义if后面的出现的语句，如果属于if语句块，则必须同一个缩进等级
- 表达式结果为Ture执行if后面的缩进的语句块

In [1]:
# if 语句练习
# 如果age小雨18岁，则打印信息“去叫家长”
age = 17
if age < 18:
    print("去叫家长把，孩子")
    print("我们不带你玩")
    print("gunqiude")
    

去叫家长把，孩子
我们不带你玩
gunqiude


In [3]:
# if 语句练习
# 如果age大于18岁，执行与if相等缩进的下一句话
age = 19
if age < 18:
    print("去叫家长把，孩子")
    print("我们不带你玩")
    print("gunqiude")

print("开始上车喽，老司机们")

开始上车喽，老司机们


### 双向分支
- if...else...语句
            if 条件表达式：
                语句1
                语句2
            else：
                语句1
                语句2

In [5]:
# input的作用是
# 1. 在屏幕上输出货号内的字符串
# 2. 在接受用户输入的内容并返回到程序
# 3. input 返回的内容一定是字符串类型

gender = input("请输入性别：")
print("你输入的性别是:{0}".format(gender))
if gender == "男":
    print("来，我们纪念一下吧，代码敲十遍")
else:
        print("来，叔叔给你糖吃")

print("开始讲for循环了")

请输入性别：女
你输入的性别是:女
来，叔叔给你糖吃
开始讲for循环了


### 多相分支

In [6]:
# 考试成绩判断
# 90分以上：输出优秀
# 80-90 : 良
# 70-80 : 中
# 60-70 : 及格
# 60以下 ：我没你这个学僧
# score 存放学生成绩
score = input("请输入学生成绩：")
# 需要把str转换成int
score = int(score)
if score>=90:
    print("A")
if score>=80 and score<90:
    print("B")
if score>=70 and score<80:
    print("C")
if score>=60 and score<70:
    print("D")
if score<60:
    print("起开，我没你这样的学生")

请输入学生成绩：54
起开，我没你这样的学生


## 多路分支
    if 条件表达式：
        语句1
        ....
    elif 条件表达式：
    
    elif 条件表达式：
    
    else:
- elif 可以有很多
- else 可选
- 多路分支只会选择一个执行
    

In [8]:
score = input("请输入学生成绩：")
# 需要把str转换成int
score = int(score)
if score>=90:
    print("A")
elif score>=80:
    print("B")
elif score>=70:
    print("C")
elif score>=60:
    print("D")
else:
    print("起开，我没你这样的学生")

请输入学生成绩：87
B


## if 语句其他：
- if 语句金额以嵌套使用，但不推荐】
- pyrhon 没有swich-ccase语句

# 循环语句
- 重复执行某些固定动作或者处理基本固定义务
- 分类
    - for循环
    - while循环
## for循环
- for 循环
        for 变量 in 序列：
            语句1
            语句2
            ...

In [9]:
# 列表就是一列数字或者其他值，一般用中括号表示
# 例如['zhangsan', 'lisi', 'wangwu', 'zhaoliu']
# 打印学生列表姓名
for name in ['zhangsan', 'lisi', 'wangwu', 'zhaoliu']:
    print(name)

zhangsan
lisi
wangwu
zhaoliu


In [10]:
# 打印学生列表姓名
# 如果是zhaoliu， 那肯定是我的最爱啊，
# 如果是别的学生，要冷酷的拒绝他
for name in ['zhangsan', 'lisi', 'wangwu', 'zhaoliu']:
    print(name)
    if name == "zhaoliu":
        print("我的最爱{0}".format(name))
    else:
        print("同学部门不约，请自重")

zhangsan
同学部门不约，请自重
lisi
同学部门不约，请自重
wangwu
同学部门不约，请自重
zhaoliu
我的最爱zhaoliu


## range介绍
- 生成一个数字序列
- 具体范围可以设定

In [12]:
# range练习
# 打印从1-10的数字
# 注意，一般在python中，如果由表示数字范围的俩个数，一般包含左边数字不包含右边数字
# randint是特列，他是左右都包括
# range函数在python2和3中有严重区别
for i in range(1,11):
    print(i)

1
2
3
4
5
6
7
8
9
10


## for-else语句
- 当for循环结束的时候，会执行else语句，else语句可选


In [13]:
# for-else语句
# 打印历表中的同学
# 如果没有在列表中，或者列表借宿了，我们需要打印提示语句，表示不再爱了
# 打印学生列表姓名
# 如果是zhaoliu， 那肯定是我的最爱啊，
# 如果是别的学生，要冷酷的拒绝他
for name in ['zhangsan', 'lisi', 'wangwu', 'zhaoliu']:
    print(name)
    if name == "zhaoliu":
        print("我的最爱{0}".format(name))
    else:
        print("同学部门不约，请自重")
else:
    print("别的都不是我的学生，不会再爱了")

zhangsan
同学部门不约，请自重
lisi
同学部门不约，请自重
wangwu
同学部门不约，请自重
zhaoliu
我的最爱zhaoliu
别的都不是我的学生，不会再爱了


## 循环之break,continue,pass
- break: 无条件结束整个循环
- continue：无条件结束本次循环。从新进入下一轮
- pass：表示略过，通常用于站位

In [14]:
# 在1-10的数字中，寻找数字7，一旦找到，打印出来，其余则什么都不做
# 在for循环中的变量表示，一般用i, k, m, n或者indx, idx, item之类
# 在python中，如果循环变量名称不重要，可以用下划线(-)代替
for i in range(1,11):
    if i == 7:
        print("我找到了7了")
        break
    else:
        print(i)
    

1
2
3
4
5
6
我找到了7了


In [15]:
# continue 语句练习
# 在数字1-10中，寻找所有偶数，找到偶数后打印偶数

for i in range(1,11):
    if i % 2 == 1:
        continue
    else:
        print("{0}是偶数".format(i))

2是偶数
4是偶数
6是偶数
8是偶数
10是偶数


In [16]:
# 版本2
for i in range(1,11):
    if i % 2 == 0:
        print("{0}是偶数".format(i))

2是偶数
4是偶数
6是偶数
8是偶数
10是偶数


In [17]:
# 版本3
for i in range(1,11):
    if i % 2 == 1:
        continue
   
    print("{0}是偶数".format(i))

2是偶数
4是偶数
6是偶数
8是偶数
10是偶数


In [18]:
# pass例子，一般用于占位
# 啥也没干，但不报错
for i in range(1,11):
    pass

# for循环练习

In [3]:
# 简单图形打印
'''
打印一下图形在输出上面
* * * * *
* * * * *
* * * * *
* * * * *
'''
print('* * * * *')
print('* * * * *')
print('* * * * *')
print('* * * * *')

* * * * *
* * * * *
* * * * *
* * * * *


In [4]:
# 更改上面代码
print('* ' * 5)
print('* ' * 5)
print('* ' * 5)
print('* ' * 5)

* * * * * 
* * * * * 
* * * * * 
* * * * * 


In [5]:
# 利用for循环更改上面案例
for i in range(0,4):
    print('* * * * * ')

* * * * * 
* * * * * 
* * * * * 
* * * * * 


In [8]:
# 利用双层for循环更改生面图案
for i in range(0,4):
    # 利用for循环打印一行星号
    for j in range(5):
        # print默认自动换行
        # 通过end参数控制，
        print('* ', end= '')
    # 利用print自动换行和利用与for循环层级相同
    print()

* * * * * 
* * * * * 
* * * * * 
* * * * * 


In [32]:
# 简单打印图形
"""
* * * * *
*       *
*       *
* * * * *
思路：
1.正常利用for循环控制打印行
2.如果是第一行和最后一行则完整打印
3.否则，判断打印列，如果是第一列或者最后一列，咋打印型号，否则打印空格
"""

# 外层循环控制行
for i in range(0,4):
    if i == 0 or i == 3:
        print('* ' * 5)
    else:
        for j in range(0,5):
            if j == 0 or j == 4 :
                print('* ',end='')
            else:
                print('  ',end='')
        print()
        

* * * * * 
*       * 
*       * 
* * * * * 


In [35]:
'''
打印三角形
*
* *
* * *
* * * *
* * * * *
'''
for i in range(5):
    # 每一行打印及几个星号跟行号相关
    print('* ' * (i + 1))
    

* 
* * 
* * * 
* * * * 
* * * * * 


In [38]:
# 另一种逻辑
for i in range(5):
    for j in range(i + 1):
        print('* ',end='')
    print()

* 
* * 
* * * 
* * * * 
* * * * * 


In [53]:
'''
打印空心三角形
*
* *
*   *
*     *
* * * * *
'''
for i in range(5):
    if i == 0 or i == 1 or i == 4:
            print('* ' * (i+1))
    else:
        for j in range(i+1):
            if j == 0 or j == (i):
                print('* ',end='')
            else:
                print('  ',end='')
        print()
            

* 
* * 
*   * 
*     * 
* * * * * 


In [57]:
'''
打印倒立三角
* * * * *
*     * 
*   *
* *
*
'''
# i-for 控制行号
# j-for控制列号
for i in range(5, 0, -1):
    for j in range(i, 0, -1):
        print('* ', end='')
    print()

* * * * * 
* * * * 
* * * 
* * 
* 


In [62]:
# 打印空三角
'''
* * * * *
*     * 
*   *
* *
*
'''
for i in range(5):
    for j in range(5-i):
        if i == 0:
            print('* ',end='')
            # continue 保证了当i=0时下面的永远不被执行
            continue
            
        if j == 0 or j == (4-i):
            print('* ',end='')
        else:
            print('  ',end='')
    print()
        


* * * * * 
*     * 
*   * 
* * 
* 


In [64]:
# 打印正三角形
'''
    *
   * *
  * * * 
 * * * *
* * * * *
'''
# i-for 控制行
# j-for 控制列

for i in range(6):
    # 总体思路是，先打印一行空格，代表每一行星星前的空格
    # 再不换行，打印型号
    for j in range(6-i):
        print(' ',end='')
    for m in range(i+1):
        print('* ',end='')
    print()
    
    

      * 
     * * 
    * * * 
   * * * * 
  * * * * * 
 * * * * * * 


# while循环
- 一个循环语句
- 表示当某条件成立的时候，就循环
- 不知道具体循环次数，但是能确定循环的成立条件的时候用while 循环
- while语法：
            
            while 条件表达式：
                语句块
                
            # 另外一种形式
            while 条件表达式：
                语句块1
            else：
                语句块2

In [2]:
# 如果说年利率是6.7%，则多少年后本钱会翻倍，年复利。
NP = 100000
year = 0
while NP <  200000:
    NP = NP * (1+0.067)
    year += 1
    print('第{0}年拿了{1}块钱'.format(year, NP))
else:
    print('我不存了')

第1年拿了106700.0块钱
第2年拿了113848.9块钱
第3年拿了121476.77629999998块钱
第4年拿了129615.72031209998块钱
第5年拿了138299.97357301068块钱
第6年拿了147566.07180240238块钱
第7年拿了157452.99861316333块钱
第8年拿了168002.34952024528块钱
第9年拿了179258.5069381017块钱
第10年拿了191268.8269029545块钱
第11年拿了204083.83830545243块钱
我不存了


# 函数
- 代码的一种组织形式
- 一个函数一般完成一项特定的功能
- 一个函数一般不要太多
- 函数使用
    - 函数需要先定义
    - 使用函数，俗称调用

In [3]:
# 定义一个函数
# 只是定义的时候不会执行
# 1.def 关键字，后跟一个空格
# 2.函数名，自己定义，起名需要遵循命名规则，不能用大驼峰
# 3. 后面括号和冒号不能省，括号内可以有参数
# 4. 函数内所有代码缩进
def func():
    print('我是一个函数')
    print('我要完成一定功能')
    print('我结素了')
# 函数的调用
func()

我是一个函数
我要完成一定功能
我结素了


### 函数的参数和返回值
- 参数：负责给函数传递一些必要的数据和信息
    - 形参：在函数定义时用到的参数没有具体值，只是一个站位的符号
    - 实参：在调用函数的时候输入的值
- 返回值：函数的执行结果
    - 使用return关键字
    - 如果没有return，默认返回一个None
    - 函数一旦执行人 return语句，则无条件返回，即结束函数的执行
    - 查找函数帮助文档

In [6]:
# 参数的定义和使用
# 参数person只是一个符号，代表的是调用的时候的某一个数据
# 调用的时候， 会用p代替函数中所有的person
def hello(person):
    print("{0}，你肿么列".format(person))
    print("sir, 你不理我我就走列")
p = "明月"
hello(p)

明月，你肿么列
sir, 你不理我我就走列


In [8]:
# return语句的基本使用
# 函数打完招呼后返回一句话
def hello(person):
    print("{0}，你肿么列".format(person))
    print("sir, 你不理我我就走列")
    
    return "我已经跟{0}打招呼了，{1}不理我".format(person, person)
p = "明月"
rst = hello(p)
print(rst)

明月，你肿么列
sir, 你不理我我就走列
我已经跟明月打招呼了，明月不理我


In [10]:
# return案例2
def hello(person):
    print("{0}，你肿么列".format(person))
    return "哈哈，我提前结速了"
    print("sir, 你不理我我就走列")
    
    return "我已经跟{0}打招呼了，{1}不理我".format(person, person)
p = "明月"
rst = hello(p)
print(rst)

明月，你肿么列
哈哈，我提前结速了


In [18]:
# 做一个99乘法表
# 自己的做法
for i in range(1,10):
    for j in range(i):
        print(str(i * (j+1)) + ' ', end = '')
    print()

1 
2 4 
3 6 9 
4 8 12 16 
5 10 15 20 25 
6 12 18 24 30 36 
7 14 21 28 35 42 49 
8 16 24 32 40 48 56 64 
9 18 27 36 45 54 63 72 81 


In [31]:
# 老师的做法
for row in range(1, 10):
    # 打印一行
    for col in range(1, row + 1):
        print(row * col, end='  ')
    print("............")
    

1  ............
2  4  ............
3  6  9  ............
4  8  12  16  ............
5  10  15  20  25  ............
6  12  18  24  30  36  ............
7  14  21  28  35  42  49  ............
8  16  24  32  40  48  56  64  ............
9  18  27  36  45  54  63  72  81  ............


In [37]:
# 定一个函数，打印一行九九乘法表
def printLine(row):
    for col in range(1, row + 1):
        print(row * col, end='  ')
    print()
for row in range(1,10):
    printLine(row)

1  
2  4  
3  6  9  
4  8  12  16  
5  10  15  20  25  
6  12  18  24  30  36  
7  14  21  28  35  42  49  
8  16  24  32  40  48  56  64  
9  18  27  36  45  54  63  72  81  


## 参数详解
- [参考资料](https://www.cnblogs.com/bingabcd/p/6671368.html)
- python 参考资料： headfirst python -> 零基础入门学习python(小甲鱼)-> 习题
- 参数分类
    - 普通参数
    - 默认参数
    - 关键字参数
    - 收集参数
- 普通参数 
    - 参加上例子
    - 定义的时候直接定义变量名
    - 调用的时候直接把变量或者值放入指定位置
        
            def 函数名（参数1，参数2， 。。。）：
                  函数体
            # 调用
            函数名（value1，value2，......)
            
            # 调用的时候，具体值参考的是未知，按位置赋值
- 默认参数
    - 形参带有默认值
    - 调用的时候，如果没有对应的形参赋值，则使用默认值
            
            def func_name(p1=vi, p2=v2.......):
                func_block
            # 调用1
            func_name()
            # 调用2(实际值跟默认值不一样）
            value1 = 100
            value2 = 200
            func_name(value1,value2)
            

In [38]:
# 默认参数示例
# 报名函数，需要知道学生性别
# 学习python 的学生基本上都是男生，所以，报名的时候没有特必然制定，我们的人物是男生
def reg(name, age, gender = 'male'):
    if gender == "male":
        print("{0} is {1}, and he is a good student".format(name, age))
    else:
        print("{0} is {1}, and she is a good student".format(name, age))
# 调用默认参数案例1
reg("mingyue", 21)
reg("xiaojing", 21, 'female')

mingyue is 21, and he is a good student
xiaojing is 21, and she is a good student


## 关键字参数开始
- 常见语法
                
                def func(p1=v1,p2=v2....):
                    func_body
                
                调用函数：
                func(p1=value1, p2=value2....)
- 比较麻烦，但也有好处
    - 不容易混淆， 一般是餐盒形参只是安位置一一对应即可，容易出错
    - 使用关键字参数，可以不考虑参数位置

In [3]:
# 关键字参数案例
def stu(name, age, addr):
    print("I am a student")
    print("我叫{0}, 我今年{1}岁了， 我住{2}".format(name, age, addr))
n = "jingjing"
a = 18
addr = "我家"
# 普通参数，只按照位置传递，容易出错
stu(a, n, addr)

I am a student
我叫18, 我今年jingjing岁了， 我住我家


In [4]:
# 关键字参数案例
def stu_key(name="No name", age=0, addr="No name"):
    print("I am a student")
    print("我叫{0}, 我今年{1}岁了， 我住{2}".format(name, age, addr))
n = "jingjing"
a = 18
addr = "我家"
# 里面参数可以换位置
# 参数不用空格，逗号后面接空格
stu_key(name=n, age=a, addr=addr)

I am a student
我叫jingjing, 我今年18岁了， 我住我家


## 收集参数
- 把没有位置，不能和定义时的参数位置相对应的参数，放入一个特定的数据结构中区
- 语法


                def func(*args):
                      func_body
                        按照list使用方式访问args得到传入的参数
                调用：
                func(p1,p2,p2,.....)
- 参数名args不是必须这么写，但是，我们推荐直接用args,约定俗成
- 参数名args前需要由星号
- 收集参数可以和其他参数共存

In [6]:
# 收集参数代码
# 函数模拟一个学生进行自我介绍，但是具体内容不清楚
# args吧他看作是一个list
def stu( *args):
    print("Hello 大家好，我自我介绍一下，简单说俩句：")
    # type函数作用是检测变量的类型
    print(type(args))
    for item in args:
        print(item)


stu("liuying", 18, "北京大通州区","wangxiaojing", "xingle")
stu("周大神")

Hello 大家好，我自我介绍一下，简单说俩句：
<class 'tuple'>
liuying
18
北京大通州区
wangxiaojing
xingle
Hello 大家好，我自我介绍一下，简单说俩句：
<class 'tuple'>
周大神


In [7]:
stu()

Hello 大家好，我自我介绍一下，简单说俩句：
<class 'tuple'>


In [8]:
# 如果使用关键字参数哥是调用，会出现问题
stu(name="liuying")

TypeError: stu() got an unexpected keyword argument 'name'

### 收集参数之关键字收集参数
- 吧关键字参数按字典格式存入收集参数
- 语法：

                        def func( **kwargs):
                                func_body
                                
                       # 调用：
                       func(p1=v1, p2= v2, p3=v3.........)
- kwargs一般约定俗成
- 调用的时候，吧多余的关键字参数放入kwargs
- 访问kwargs需要按字典格式访问

In [13]:
# 收集参数案例
# 自我介绍
# 调用时使用关键字参数调用
def stu( **kwargs):
    print("Hello 大家好，我先自我介绍一下:")
    print(type(kwargs))
    # 对于字典的访问，python2和python3的区别
    for k,v in kwargs.items():
        print(k, "---", v)

stu(name="liuying", age=19, addr="北京大通州区", love="wangxiaojing", work="teacher")
print("********")
stu(name="zhoudashen")
    

Hello 大家好，我先自我介绍一下:
<class 'dict'>
name --- liuying
age --- 19
addr --- 北京大通州区
love --- wangxiaojing
work --- teacher
********
Hello 大家好，我先自我介绍一下:
<class 'dict'>
name --- zhoudashen


### 收集参数混合调用的顺序问题
- 收集参数，关键字参数，普通参数可以混合使用
- 使用规则，普通参数和关键字参数优先处理
- 定义的时候一般宅普通参数，关键字参数，收集参数tuple，收集参数dict

In [31]:
# 收集参数混合使用案例
# stu 模拟一个学生的自我介绍
def stu(name, age, *args, hobby="没有", **kwargs):
    print("Hello 大家好")
    print("我叫{0},我今年{1}大了。".format(name, age))
    if hobby == "没有":
        print("对不起，我没有爱好")
    else:
        print("我的爱好是{0}".format(hobby))
    for i in args:
        print(i)   
   
    
    
    for k, v in kwargs.items():
        print(k, "---", v)
    print("*" * 20)

# 开始调用函数
name = "liuying"
age = 19
# 调用的不同格式
stu(name, age)
stu(name, age, hobby = "游泳")
stu(name, age, "wangxiaojing", "liushitou", hobby="游泳", hobby2="烹饪",hobby3="跟别人聊天")

Hello 大家好
我叫liuying,我今年19大了。
对不起，我没有爱好
********************
Hello 大家好
我叫liuying,我今年19大了。
我的爱好是游泳
********************
Hello 大家好
我叫liuying,我今年19大了。
我的爱好是游泳
wangxiaojing
liushitou
hobby2 --- 烹饪
hobby3 --- 跟别人聊天
********************


### 收集参数的解包问题
- 吧参数放入list或者字典中，直接吧list/dict中的树枝放入参数重
- 语法：参看案例

In [33]:
# 收集参数的解包问题
def stu(*args):
    print("hahahaah")
    # n 用来表示循环次数
    # 主要用来调试
    n = 0
    for i in args:
        print(type(i))
        print(n)
        n += 1
        print(i)
stu("liuying", "liuxiaoying", 19, 200)
l = list()
l.append("liuying")
l.append(20)
l.append(230)
stu(l)

hahahaah
<class 'str'>
0
liuying
<class 'str'>
1
liuxiaoying
<class 'int'>
2
19
<class 'int'>
3
200
hahahaah
<class 'list'>
0
['liuying', 20, 230]


In [36]:
# 收集参数的解包问题2
def stu(*args):
    print("hahahaah")
    # n 用来表示循环次数
    # 主要用来调试
    n = 0
    for i in args:
        print(type(i))
        print(n)
        n += 1
        print(i)
l = ["liuying", 19, 23, "wangxiaojing"]
stu(l)
# 此时，args的表示形式是字典里有一个元素，即args =  (["liuying", 19, 23, "wangxiaojing"])
# 很显然，违背最初想法

# 此时的调用，我们要解包加个星号
stu(*l)

hahahaah
<class 'list'>
0
['liuying', 19, 23, 'wangxiaojing']
hahahaah
<class 'str'>
0
liuying
<class 'int'>
1
19
<class 'int'>
2
23
<class 'str'>
3
wangxiaojing


### 同理，dict类型收集参数一样可以解包，但是
- 对dict类型进行解包需要俩个**

## 返回值
- 函数和过程的区别
    - 有无返回值
- 需要用return显示返回值内容，
- 如果没有返回，则默认返回None
- 推荐写法，无论有没有返回值，最后都以return结速了

In [39]:
# 返回值示例
def finc_1():
    print("有返回值呀")
    return 1

def finc_2():
    print("没有返回值")
    return None

f1 = finc_1()
print(f1)

有返回值呀
1


In [40]:
f2 = finc_2()
print(f2)

没有返回值
None


# 函数文档
- 函数的文档的作用是对当前函数提供使用相关的参考信息
- 文档的写法：
    - 在函数内部开始的第一行使用3字符串定义符
    - 一般具有特定合适
    - 参看案例
- 文档查看
    - 使用help函数，形如help(func)
    - 使用doc

In [44]:
# 文档案例
# 函数stu是模拟一个学生的自我介绍的内容
def stu(name, age, *args):
    '''
    这是文档
    这是文档
    这是文档
    '''
    print("This is hanshu stu")

In [45]:
# 查看函数文档
help(stu)
# 使用doc
stu.__doc__

Help on function stu in module __main__:

stu(name, age, *args)
    这是文档
    这是文档
    这是文档



'\n    这是文档\n    这是文档\n    这是文档\n    '

# 变量作用域
- 变量有作用范围限制
- 分类：按照作用域分类
     - 全局(global):在函数外部定义
     - 局部(local):在函数内部定义
- 变量的作用范围：
    - 全局变量：在整个全局范围都有效
    - 全局变量在局部可以使用（即函数内部可以可以发表范围函数外部定义的变量）
    - 局部变量在局部范围可以使用
    - 局部变量在全局范围无法使用
- LEGB原则
    - L（local)局部作用域
    - E（enclosing function local）外部嵌套函数作用域
    - G(Global module）函数定义所在模块作用域
    - B(Buildin）：python内置模块的作用域
    可以从更小的范围访问小范围，但不能从外部访问函数内
    如果出现函数内外都出现的同样变量，函数内部访问函数外部的不行
    函数外不能访问函数内部变量
    

In [6]:
# 例子
a1 = 100
def fun():
    print(a1)
    print("I am in fun")
    a2 = 99
    
    print(a2)
print(a1)
fun()
print(a2)

100
100
I am in fun
99


NameError: name 'a2' is not defined

## 提升局部变量为全局变量
- 使用global
- 案例如下

In [12]:

def fun():
    global b1
    b1 = 100
    print(b1)
    print("I am in fun")
    # a2 的作用范围是fun
    b2 = 99
    print(b2)
print(b1)
fun()

100
100
I am in fun
99


# globals, locals函数
- 可以通过globals和locals显示出局部变量和全局变量
- 参看以下案例

In [13]:
# globals 和 locals
a = 1
b = 2

def fun(c,d):
    e = 111
    print('Locals={0}'.format(locals()))
    print('Globals={0}'.format(globals()))
fun(100, 200)

Locals={'c': 100, 'd': 200, 'e': 111}
Globals={'__name__': '__main__', '__doc__': 'Automatically created module for IPython interactive environment', '__package__': None, '__loader__': None, '__spec__': None, '__builtin__': <module 'builtins' (built-in)>, '__builtins__': <module 'builtins' (built-in)>, '_ih': ['', '# 例子\na1 = 100\ndef fun():\n    print(a)\n    print("I am in fun")\n    a2 = 99\n    \n    print(a2)\nprint(a1)', '# 例子\na1 = 100\ndef fun():\n    print(a)\n    print("I am in fun")\n    a2 = 99\n    \n    print(a2)\nprint(a1)\nprint(a2)', '# 例子\na1 = 100\ndef fun():\n    print(a)\n    print("I am in fun")\n    a2 = 99\n    \n    print(a2)\nprint(a1)\nfun()\nprint(a2)', '# 例子\na1 = 100\ndef fun():\n    print(a)\n    print("I am in fun")\n    a2 = 99\n    \n    print(a2)\nprint(a1)\nfun()', '# 例子\na1 = 100\ndef fun():\n    print(a1)\n    print("I am in fun")\n    a2 = 99\n    \n    print(a2)\nprint(a1)\nfun()', '# 例子\na1 = 100\ndef fun():\n    print(a1)\n    print("I am in fu

# eval()函数
- 把一个字符串当成一个表达式来执行，返回表达式执行后的结果
- 语法：

                    eval(string_code, globals=None, locals=None)
# exec()函数
- 跟eval功能类事，但是，不返回结果
- 语法：

                         exec（string_code, globals=None, locals=None)
                
                 
                    

In [16]:
x = 100
y = 200
# 执行x+y
# z = x + y
z1 = x + y
z2 = eval("x+y")
z3 = exec("x+y")
z4 = exec('print("x+y", x+y)')
print(z1)
print(z2)
print(z3)
print(z4)

x+y 300
300
300
None
None


# 递归函数
- 函数直接或者间接调用自身
- 优点：间接，理解容易
- 缺点：对递归深度有限制，消耗资源大，python有递归层数限制，超过限制报错
- 在写递归程序的时候，一定要注意结束层数

In [19]:
# 递归调用深度限制代码
x = 0
def fun():
    global x
    x += 1
    print(x)
    # 函数调用自己
    fun()
# 调用函数
fun()

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


2325
2326
2327
2328
2329
2330
2331
2332
2333
2334
2335
2336
2337
2338
2339
2340
2341
2342
2343
2344
2345
2346
2347
2348
2349
2350
2351
2352
2353
2354
2355
2356
2357
2358
2359
2360
2361
2362
2363
2364
2365
2366
2367
2368
2369
2370
2371
2372
2373
2374
2375
2376
2377
2378
2379
2380
2381
2382
2383
2384
2385
2386
2387
2388
2389
2390
2391
2392
2393
2394
2395
2396
2397
2398
2399
2400
2401
2402
2403
2404
2405
2406
2407
2408
2409
2410
2411
2412
2413
2414
2415
2416
2417
2418
2419
2420
2421
2422
2423
2424
2425
2426
2427
2428
2429
2430
2431
2432
2433
2434
2435
2436
2437
2438
2439
2440
2441
2442
2443
2444
2445
2446
2447
2448
2449
2450
2451
2452
2453
2454
2455
2456
2457
2458
2459
2460
2461
2462
2463
2464
2465
2466
2467
2468
2469
2470
2471
2472
2473
2474
2475
2476
2477
2478
2479
2480
2481
2482
2483
2484
2485
2486
2487
2488
2489
2490
2491
2492
2493
2494
2495
2496
2497
2498
2499
2500
2501
2502
2503
2504
2505
2506
2507
2508
2509
2510
2511
2512
2513
2514
2515
2516
2517
2518
2519
2520
2521
2522
2523
2524


RecursionError: maximum recursion depth exceeded while calling a Python object

In [1]:
# 斐波那契数列
# 一列数字，第一个数字1，第二个也是1，从地个开始，每一个数字的值等于前俩个数字出现的值的和
# 数学公式为： f(1)=1, f(2)= 1, f(n-1)+f(n-2)
# 例如：1，1，2，3，5，8，13。。。。。

# 第n个数字的斐波那契数列的值
def fib(n):
    if n == 1:
        return 1
    
    if n == 2:
        return 1
    if n <= 0:
        return None
    
    return fib(n-1) + fib(n-2)
print(fib(5))
    

5


# 内置数据结构（变量类型）
- list
- set
- dict
- tuple
## list(列表）
- 一组有顺序的数据的组合
- 创建列表
    - 空列表

In [6]:
# 创建空列表
l1 = []
print(type(l1))
print(l1)
# 创建带值的列表
l2 = [100]
print(type(l2))
print(l2)

l3 = [2,3,1,4,5,3,5]
print(type(l3))
print(l3)

l4 = list()
print(type(l4))
print(l4)



<class 'list'>
[]
<class 'list'>
[100]
<class 'list'>
[2, 3, 1, 4, 5, 3, 5]
<class 'list'>
[]


## 列表常用操作
- 访问
    - 使用下标操作（索引）
    - 列表的位置是从0开始
- 分片操作
    - 对列表进行任意一段的截取
    - l[:]
    

In [10]:
# 下标访问列表
l = [3,2,1,4,6,3,2]
print(l[3])
# 分片操作,包左不包右
print(l[1:4])
# 下标值可以为空
print(l[:])


4
[2, 1, 4]
[3, 2, 1, 4, 6, 3, 2]


In [11]:
# 分片操作可以控制增长幅度，默认增长幅度为一
print(l[1:6:1])

[2, 1, 4, 6, 3]


In [14]:
# 打印下标1开始的数据，每次隔一个
print(l[1:6:2])
# 下标可以超出范围，超出后不再考虑多余下标内容
print(l[2:10])
# 下标值，增长幅度可以为负数
# 为负数，表明顺序是从右往左
# 规定：数组最后一个数字的下标是-1

[2, 4, 3]
[3, 2]


## 分片操作是生成一个新的list
- 内置函数id,负责显示一个变量或者数据的唯一确定编号

In [17]:
# id函数剧烈
a = 100
b = 200
print(id(a))
print(id(b))
# 如果a跟c只想一份数据，则更改a的值同样也会更改c的值
# 但是结果并非如此，为什么？
c = a 
a = 101
print(id(c))
print(a)
print(c)

4300628128
4300631328
4300628128
101
100


In [21]:
# 通过id可以直接判断出分片是重新生成可一份数据还是使用的同一份数据
l = [3,4,56,76,32,21,43,5]
ll = l[:]
lll = ll
# 如果俩个id值一样，则表明分片产生的列表是使用的同一地址同一份数据
print(id(l))
print(id(ll))
print(id(lll))
# 通过id知道，ll和lll是同一份数据，验证代码如下
l[1] = 100
print(l)
print(ll)

ll[1] = 100
print(ll)
print(lll)

4334596488
4333947976
4333947976
[3, 100, 56, 76, 32, 21, 43, 5]
[3, 4, 56, 76, 32, 21, 43, 5]
[3, 100, 56, 76, 32, 21, 43, 5]
[3, 100, 56, 76, 32, 21, 43, 5]


# 汉诺塔问题
- 规则1，每次只能移动一个盘子，2，任何一次移动，3个塔的的状态必须是大盘子在下，小盘子在上
- 移动，n =1，a->c
- n=2: a->b a->c b->c
- n=3:变成n = 2
    - 把a上的俩个盘子，通过c移动到b上去，调用递归实现
    - 把a上的剩下的最大的盘子移动到c上，a->c
    - 吧b上俩个盘子，借助a，移动到c上去，调用递归
- n=n:
    - 把a上的n-1个盘子，借助c，移动到b上去，调用递归
    - 吧a上的最大盘子，也是唯一一个，移动号到c上，a->c
    - 把b上n-1个盘子，借助a，移动c上去，调用递归

In [2]:
def hano(n, a, b, c):
    """
    汉诺塔的递归实现
    n:  代表几个盘子
    a:  代表第一个塔
    b：代表第二个塔
    c：代表第三个塔
    遵守汉诺塔规则
    """
    if n == 1:
        print(a, "-->", c)
        return None
    """
    if n == 2:
        print(a, "-->", b)
        print(a, "-->", c)
        print(b, "-->", c)
        return None
    """
    # 把n-1个盘子，从a塔借助c塔，挪到b上去
    hano(n-1, a, c, b)
    print(a, "-->", c)
    # 把n-1和盘子，从b塔，借助于a塔，挪到c上
    hano(n-1, b, a, c)
a = "A"
b = "B"
c = "C"
hano(3, a, b, c)

A --> C
A --> B
C --> B
A --> C
B --> A
B --> C
A --> C


# List(列表）
- del:删除命令

In [3]:
# del 删除
a = [1, 2, 3, 4, 5, 6]
del a[2]
print(a)

[1, 2, 4, 5, 6]


In [4]:
# del 删除
# 如归使用del之后，id的值和删除前不一样，则说明删除生成了一个新的list
a = [1, 2, 3, 4, 5, 6]
print(id(a))
del a[2]
print(a)
print(id(a))


4452247688
[1, 2, 4, 5, 6]
4452247688


In [5]:
# del一个变量之后不能再使用此变量
del a
print(a)

NameError: name 'a' is not defined

### 列表相加
- 使用加号连接俩个列表

In [6]:
a = [1, 2, 3, 4, 5]
b = [6, 7, 8, 9, 10]
d = ['a', 'b', 'c']
c = a + b + d
print(c)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 'a', 'b', 'c']


In [7]:
# 使用乘号操作列表
# 列表直接跟一个整数相乘
# 相当于把n个列表接在一起
a = [1, 2, 3, 4, 5]
b = a * 10
print(b)

[1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5]


In [9]:
# 成员资格运算
# 就是判断一个元素是否在list里边
a = [1, 2, 3, 4, 5, 6]
b = 8
# c 的值是一个布尔值
c = b in a
print(c)
b = 4
print(b in a)

False
True


In [10]:
# not in 
a = [1, 2, 3, 4, 5]
b = 9
print(b not in a)

True


# 链表的遍历
- for
- while

In [8]:
# for in list
a = [1,2,3,4,5]
# 挨个打印出来
for i in a:
    print(i)

1
2
3
4
5


In [13]:
b = ["i love wangxiaojing"]
for i in b:
    print(i)

i love wangxiaojing


In [14]:
# range
for i in range(1,10):
    print(i)
    print(type(range))

1
<class 'type'>
2
<class 'type'>
3
<class 'type'>
4
<class 'type'>
5
<class 'type'>
6
<class 'type'>
7
<class 'type'>
8
<class 'type'>
9
<class 'type'>


In [4]:
# while 循环访问list
# 一般情况下不用while
a = [1,2,3,4,5,6]
length = len(a)
# index 表示的是list的下标
index = 0
while index < length:
    print(a[index])
    index += 1
    

1
2
3
4
5
6


In [9]:
# 双层列表循环
# a为嵌套列表，或双层列表
a = [["one", 1], ["two", 2], ["three", 3]]
for k,v in a:
    print(k, "--", v)
    

one -- 1
two -- 2
three -- 3


In [10]:
# 双层列表循环
# a为嵌套列表，或双层列表
a = [["one", 1], ["two", 2], ["three", 3，4，5，6，7]]
for k,v in a:
    print(k, "--", v)
    

SyntaxError: invalid character in identifier (<ipython-input-10-e12808e9ae08>, line 3)

In [12]:
# 双层列表循环
# a为嵌套列表，或双层列表
a = [["one",1,"eins"], ["two",2,"zwei"], ["three",3,"drei"]]
for k,v,w in a:
    print(k, "--", v,"--",w)
    

one -- 1 -- eins
two -- 2 -- zwei
three -- 3 -- drei


# 列表内涵：list content
- 通过简单方法仓做列表

In [13]:
# for 创建
a = ["a", "b", "c"]
# 用list a创建一个list b
# 下面代码的含义是，对于所有a中的元素，逐个放入新的列表b中
b = [i for i in a]
print(b)

['a', 'b', 'c']


In [14]:
# 对a中的所有元素乘以10，生成一个新的list
a = [1,2,3,4,5]
# 用list a 创建一个list b
# 下面代码的含义是，对于所有a中的元素，逐个放入新的列表b中
b = [i * 10 for i in a ]
print(b)

[10, 20, 30, 40, 50]


In [15]:
# 还可以过滤掉原来list中的内容放入新list
# 比如原油列表a， 需要把所有a中有偶数生成新的列表b

a = [x for x in range(1,35)]# 生成从1到34的一个列表
# 把a中所有偶数生成一个新的列b
b = [m for m in a if m % 2 == 0]
print(b)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34]


In [22]:
# 列表生成式也可以嵌套
# 由俩个列表a,b
a = [i for i in range(1,4)] # 生成list a
print(a)

b = [i for i in range(100,400) if i % 100 == 0]
print(b)
# 列表生成时可以嵌套
c = [m + n for m in a for n in b]
print(c)

for m in a:
    for n in b:
        print(m+n, end="  ")
print()

c = [m+n for m in a for n in b if m+n <250]
print(c)

[1, 2, 3]
[100, 200, 300]
[101, 201, 301, 102, 202, 302, 103, 203, 303]
101  201  301  102  202  302  103  203  303  
[101, 201, 102, 202, 103, 203]


# 关于列表的常见函数

In [25]:
# len:求列表长度
a = [x for x in range(1,100)]
print(len(a))

# max:求列表中的最大值
# min: 同理
print(max(a))

b = ['man', 'film', 'python']
print(max(b))

99
99
python


In [27]:
# list：将其他格式数据转换成list
a = [1, 2, 3]
print(list(a))

[1, 2, 3]


In [28]:
s = "woaiwangxiaojing"
print(list(s))

['w', 'o', 'a', 'i', 'w', 'a', 'n', 'g', 'x', 'i', 'a', 'o', 'j', 'i', 'n', 'g']


In [29]:
print(list(range(12, 19)))
# 被转换的格式必须时可迭代的

[12, 13, 14, 15, 16, 17, 18]


In [1]:
# 一个同学的问题
# 传值和传地址的区别，值和列表的区别
def a(n):
    n[2] = 300
    print(n)
    return None

def b(n):
    n += 100
    print(n)
    return None

an = [1, 2, 3, 4, 5, 6]
bn = 9
print(an)
a(an)
print(an)

print(bn)
b(bn)
print(bn)

[1, 2, 3, 4, 5, 6]
[1, 2, 300, 4, 5, 6]
[1, 2, 300, 4, 5, 6]
9
109
9


# 关于列表的函数

In [3]:
l = ['a', 'i love wangxiaojing', 45, 766, 5+4j]
l

['a', 'i love wangxiaojing', 45, 766, (5+4j)]

In [4]:
# append 插入一个内容,在末尾追加
a = [i for i in range(1,5)]
print(a)
a.append(100)
print(a)
# insert:指定位置插入
# insert(index, data)
a.insert(3, 666)
print(a)

[1, 2, 3, 4]
[1, 2, 3, 4, 100]
[1, 2, 3, 666, 4, 100]


In [6]:
# 删除
# del 删除第几个元素
# pop， 从对位拿出一个元素，即把最后一个元素取出来
# pop 把拿出来的元素放到别的变量中
last_ele = a.pop()
print(last_ele)
print(a)

4
[1, 2, 3, 666]


In [11]:
# remove：在列表中删除指定的值的元素
# 如果被删除的值没在list中，则报错
# 即， 删除list指定值的操作应该使用try。。。excepty语句，或者进行先行判断
# if x in list:
# list.remove(x)
a.insert(4,666)
print(id(a))
print(a)

a.remove(666)
print(a)
print(id(a))
# 输出俩个id值一样，说明，remove操作时子啊远list直接操作

4401059080
[1, 2, 3, 666]
[1, 2, 3]
4401059080


In [13]:
# clear:清空
print(a)
print(id(a))
a.clear()
print(id(a))


[1, 2, 3]
4401059080
4401059080


In [15]:
# 如果不需要列表地址保持不变，则清空列表可以使用以下方式
# a = list()
# a = []
# reverse:翻转列表内容，原地反转
a = [1,2,3,4,5]
print(a)
print(id(a))

a.reverse()

print(a)
print(id(a))

[1, 2, 3, 4, 5]
4402405000
[5, 4, 3, 2, 1]
4402405000


In [16]:
# extend:翻转列表内容，原地反转
a = [1,2,3,4,5]
print(a)
print(id(a))

a.reverse()

print(a)
print(id(a))

[1, 2, 3, 4, 5]
4401058120
[5, 4, 3, 2, 1]
4401058120


In [19]:
# count:查找列表中指定值或元素的个数
print(a)
a.append(8)
a.insert(3,8)
a_len = a.count(8)
print(a)
print(a_len)

[5, 4, 3, 8, 8, 2, 1, 8, 8]
[5, 4, 3, 8, 8, 8, 2, 1, 8, 8, 8]
6


In [24]:
# copy:拷贝，浅拷贝
# 列表类型变量赋值示例
a = [1,2,3,4,5,666]
print(a)
# list类型，简单赋值操作，是传地址
b = a
b[3] = 777
print(a)
print(b)

print('*' * 30)

# 为了解决上述传地址问题，list赋值需要采用copy函数
a = [1,2,3,4]

b = a.copy()
b[3] = 777
print(a)
print(id(a))
print(b)
print(id(b))

[1, 2, 3, 4, 5, 666]
[1, 2, 3, 777, 5, 666]
[1, 2, 3, 777, 5, 666]
******************************
[1, 2, 3, 4]
4402553032
[1, 2, 3, 777]
4402554056


In [25]:
# 深拷贝跟深拷贝的区别
a = [1,2,3,[10,20,30]]
b = a.copy()
print(id(a))
print(id(b))
print(id(a[3]))
print(id(b[3]))
a[3][2] = 666
print(a)
print(b)

# 后面的【10，20，30】只是拷贝了一个地址，copy是一个浅拷贝，只拷贝一层
# 深拷贝需要使用特定工具

4402553160
4402553992
4402553224
4402553224
[1, 2, 3, [10, 20, 666]]
[1, 2, 3, [10, 20, 666]]


# 元祖- tuple
- 元祖可以看成一个不可更改的list
## 元祖创建


In [30]:
# 创建空元祖
t = ()
print(type(t))
# 创建一个只有一个值的元祖
t = (1,)
print(type(t))
print(t)
t = 1,
print(type(t))
print(t)

# 创建多个值
t = (1,2,3,4,5)
print(type(t))
print(t)

t = 1,2,3,4,5
print(type(t))
print(t)

l = [1,2,3,4,5]
t = tuple(l)
print(type(t))
print(t)




<class 'tuple'>
<class 'tuple'>
(1,)
<class 'tuple'>
(1,)
<class 'tuple'>
(1, 2, 3, 4, 5)
<class 'tuple'>
(1, 2, 3, 4, 5)
<class 'tuple'>
(1, 2, 3, 4, 5)


## 元祖的特性
- 是虚列表，有序
- 原则数据可以访问，不能修改
- 元祖数据可以是任何了形
- 总之，list的特性除了不能修改都有
- 也就意味着，list基友的一些操作，不如索引，分片，序列想家，相乘，成员呢资格操作等，一摸 - 一样

In [31]:
# 索引操作
t = (1,2,3,4,5)
print(t[4])

5


In [33]:
# 超标错误
print(t[12])

IndexError: tuple index out of range

In [36]:
t = (1,2,3,4,5,6)
t1 = t[1::2]
print(id(t))
print(id(t1))
print(t1)

# 切片可以超标

4381183240
4403134400
(2, 4, 6)


In [39]:
# 序列相加
t1 = (1,2,3)
t2 = (5,6,7)

# 传地址操作
print(t1)
print(id(t1))
t1 += t2
print(t1)
print(id(t1))

# 以上操作，类似于
t1 = (1,2,3)
t1 = (2,3,4)

# tuple的不可修改，值的是内容的不可修改

t1[1] = 100
# 会报错

(1, 2, 3)
4403174784
(1, 2, 3, 5, 6, 7)
4401381288


TypeError: 'tuple' object does not support item assignment

In [40]:
# 元祖相乘
t = (1,2,3)
t = t * 3
print(t)

(1, 2, 3, 1, 2, 3, 1, 2, 3)


In [42]:
# 成员检测
t = (1,2,3)
if 2 in t:
    print("yes")
else:
    print("No")
    

yes


In [44]:
# 元祖遍历，一般用for
# 1，单层元祖遍历
t = (1,2,3,'wangxiaojing', 'i', 'love')
for i in t:
    print(i, end = ' ')

1 2 3 wangxiaojing i love 

In [45]:
# 2双层元祖的遍历
t = ((1,2,3),(2,3,4),('wangxiaojing', 'i', 'love'))
# 对以上元祖的遍历，可以如下
# 1

for i in t:
    print(i)

(1, 2, 3)
(2, 3, 4)
('wangxiaojing', 'i', 'love')


In [46]:
for k,m,n in t:
    print(k, '--',m,'--',n)

1 -- 2 -- 3
2 -- 3 -- 4
wangxiaojing -- i -- love


# 习题课-str list dict
- str
- list
- dict
- set

In [51]:
# 字符串的简单操作
# 连接操作
s1 = "i love "
s2 = "wangxiaojing"
s3 = s1 + s2
print(s3)
print(s1 + ' ' + s2)

i love wangxiaojing
i love  wangxiaojing


In [53]:
# 字符串的乘法
s = "i love wangxiaojing"
s2 = s * 10
print(s2)

i love wangxiaojingi love wangxiaojingi love wangxiaojingi love wangxiaojingi love wangxiaojingi love wangxiaojingi love wangxiaojingi love wangxiaojingi love wangxiaojingi love wangxiaojing


In [54]:
# 字符串当成列表
s = " i love wang xiangjing"
print(s[1])

i


In [55]:
print(s[-1])

g


In [56]:
s = "123456789"
print(s[2:6])

3456


In [57]:
print(s[:])

123456789


In [4]:
# 切片返回的是一个全新的字符串吗
# 主要看切的是
print(id(s))
print(id(s[:]))
print(id(s[:5]))

# 切片返回的是一个全新的字符串吗》
# 如果切片去一部分，则返回全新字符串
# 如果取完整切片，可能返回内容指向同一字符串

s = "i love wangxiaojing and zhangxiaojing"
s1 = s[:]
# 不能直接对字符串进行下标元素赋值，否则会导致错误
s1 =[1] = "A"
print(s)
print(s1)


SyntaxError: can't assign to literal (<ipython-input-4-755f680e71b1>, line 14)

In [7]:
# 用实验证明部分切片和完整切片的区别
l = [1,2,3,4,5,6,7,8,9]
l1 = l[:2]
print(id(l))
print(id(l1))

4391019016
4392533768


Help on class str in module builtins:

class str(object)
 |  str(object='') -> str
 |  str(bytes_or_buffer[, encoding[, errors]]) -> str
 |  
 |  Create a new string object from the given object. If encoding or
 |  errors is specified, then the object must expose a data buffer
 |  that will be decoded using the given encoding and error handler.
 |  Otherwise, returns the result of object.__str__() (if defined)
 |  or repr(object).
 |  encoding defaults to sys.getdefaultencoding().
 |  errors defaults to 'strict'.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(self, format_spec, /)
 |      Return a formatted version of the string as described by format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  

# 关于元祖的函数
- 查看以下代码
- 基本跟list通用

In [3]:
# count: 计算指定数据出现的次数
t = (2,1,2,3,45,1,2,1)
print(t.count(2))

# index 元素在元祖中的索引位置
print(t.index(45))

# 如果出现多个元素，则返回第一个
print(t.index(1))

3
4
1


# 元祖变量交换法
- 俩个变量交换值

In [ ]:
# 俩个变量交换值
a = 1
b = 3

print(a)
print(b)
print("*" * 20)
# java程序员会这么写：
c = a
a = b
b = c
print(a)
print(b)

print("*" * 20)
# python的写法
a,b = b,a
print(a)
print(b)

# 集合-set
- 集合是高中数学中的一个概念
- 一堆确定的无序的唯一的数据，集合中每一个数据成为一个元素



In [8]:
# 集合的定义
s = set()
print(type(s))
print(s)


# 此时大括号里一定要有值
s = {1,2,3,4,5,6}
print(type(s))
print(s)

<class 'set'>
set()
<class 'set'>
{1, 2, 3, 4, 5, 6}


In [7]:
# 如果只是用大括号定义，则定义的是一个dict类型
d = {}
print(type(d))3
print(d)

<class 'dict'>


# 集合的特征
- 集合的数据无序， 即无法索引和分片
- 集合内部数据具有唯一性，可以用来排出重复数据
- 集合内的数据，str, int, float, tuple, 冰冻集合等，即内部只能放置可哈希数据

# 集合序列的操作

In [13]:
# 成员检测
# in ,not in
s = {4,5,"i",'love',"wangxiaojing"}
print(s)
if "love" in s:
    print("aiya")

{4, 5, 'wangxiaojing', 'i', 'love'}
aiya


# 集合遍历操作

In [14]:
# for 循环
s = {4,5,"i",'love',"wangxiaojing"}
for i in s:
    print(i, end=" ")

4 5 wangxiaojing i love 

In [15]:
# 带有元祖的集合遍历
s = {(1,2,3), ("i","love","wangxiaojing"),(4,5,6)}
for k,m,n in s:
    print(k,m,n)

i love wangxiaojing
4 5 6
1 2 3


In [16]:
for k in s:
    print(k)

('i', 'love', 'wangxiaojing')
(4, 5, 6)
(1, 2, 3)


# 集合的内涵


In [2]:
# 普通的集合
# 以下集合在初始化后自动过滤掉重复元素
s = {23,223,545,3,1,2,3,4,3,2,3,1,2,4,3}
print(s)
# 普通集合内涵
ss = {i for i in s}
print(ss)

{545, 1, 3, 2, 4, 23, 223}
{545, 2, 3, 4, 1, 23, 223}


In [3]:
# 带条件的集合内涵
sss = {i for i in s if i % 2 == 0}
print(sss)

{2, 4}


In [6]:
# 多循环的集合内涵
s1 = {1,2,3,4}
s2 = {"i", "love", "wangxiaojing"}
s = {m*n for m in s1 for n in s2}
print(s)
s = {m*n for m in s1 for n in s2 if n == 2}
print(s)

{'i', 'ii', 'lovelove', 'iii', 'lovelovelove', 'wangxiaojingwangxiaojingwangxiaojing', 'wangxiaojingwangxiaojing', 'wangxiaojing', 'iiii', 'wangxiaojingwangxiaojingwangxiaojingwangxiaojing', 'love', 'lovelovelovelove'}
set()


In [8]:
# len, max, min :跟其他函数基本一致
# 生成一个集合
l = [1,2,3,4,2,23,1,2,3,4]
s = set(l)
print(s)

{1, 2, 3, 4, 23}


In [10]:
# add: 向集合内部添加元素
s = {1}
s.add(223)
print(s)

{1, 223}


In [11]:
# clear
s = {1,2,3,4,5}
print(id(s))
s.clear()
print(id(s))
# 结果表明clear原地清空数据

4390768200
4390768200


In [12]:
# copy：拷贝
# remove：移除指定值，没有的话报错
# discard:移除集合中指定的值，没有的话不报错
# pop随机移除一个元素
s = {1,2,3,4,5,6,7}
d = s.pop()
print(d)
print(s)

1
{2, 3, 4, 5, 6, 7}


In [13]:
# 集合函数
# intersection:交集
# difference: 差集
# union: 并集
# issubset:检查一个集合是否为另一个子集
# issuperset:检查一个集合是否为另一个集合的超集
s1 = {1,2,3,4,5,6}
s2 = {5,6,7,8,9}

s_1 = s1.intersection(s2)
print(s_1)

{5, 6}


In [14]:
s_2 = s1.difference(s2)
print(s_2)

{1, 2, 3, 4}


In [15]:
s_3 = s1.issubset(s2)
print(s_3)

False


In [17]:
# 集合的数学操作
s1 = {1,2,3,4,5,6}
s2 = {5,6,7,8,9}

s_1 = s1 - s2
print(s_1)

{1, 2, 3, 4}


# frozen set:冰冻集合
- 就是不能进行任何修改的集合
- frozen是一种特殊集合

In [20]:
# 创建
s = frozenset()
print(type(s))
print(s)

<class 'frozenset'>
frozenset()


# dict字典
- 字典是一种组合数据，没有顺序的组合数据，数据以键值对形式出现

In [24]:
# 字典的创建
d = {}
print(d)
# 创建空字典
d = dict()
print(d)
# 创建有值的字典，每一组数据用冒号隔开，每一对键值对用逗号隔开
d = {"one":1, "two":2, "three":3}
print(d)

# 用dict创建有内容字典2
# 利用关键字参数
d = dict(one=1, two=2, three=3)
print(d)

#
d = dict([("one",1),("two",2),("three",3)])
print(d)

{}
{}
{'one': 1, 'two': 2, 'three': 3}
{'one': 1, 'two': 2, 'three': 3}
{'one': 1, 'two': 2, 'three': 3}


# 字典的特征
- 字典是序列类型，但是是无序列，所以没有分片和索引
- 字典中的数据每个都有键值对，即kv对
    - key:必须是可哈希的值，比如int,string,float,tuple,但是list,set,dict不行
    - value :任何值

In [29]:
# 字典常见操作
# 访问数据
d = {"one":1, "two":2, "three":3}
# 注意访问格式
# 中括号里面是键值
print(d["one"])

d["one"] = "eins"
print(d)
# 删除某个值
del d["one"]
print(d)
# 成员检测， in, not in
d = {"one":1, "two":2, "three":3}
# 成员检测，其实只是检测key
if 2 in d:
    print("value")
    
if "two" in d:
    print("key")
    
if ("two",2) in d:
    print("kv")

1
{'one': 'eins', 'two': 2, 'three': 3}
{'two': 2, 'three': 3}
key


In [33]:
# 遍历在python2和python3区别比较大，代码不通用
# 按key来使用for循环
d = {"one":1, "two":2, "three":3}
# 使用for循环，直接用key访问
for k in d:
    print(k, d[k])
# 上述代码可以改写成如下
for k in d.keys():
    print(k,d[k])
# 只访问字典的值
for v in d.values():
    print(v)
# 注意以下特殊用法
for k,v in d.items():
    print(k,'--',v)
    

one 1
two 2
three 3
one 1
two 2
three 3
1
2
3
one -- 1
two -- 2
three -- 3


In [34]:
# 字典的生成式
d = {"one":1, "two":2, "three":3}
# 常规字典生成式
dd = {k:v for k,v in d.items()}
print(dd)

{'one': 1, 'two': 2, 'three': 3}


In [35]:
# 加限制条件的字典生成式
dd = {k:v for k,v in d.items() if v % 2 == 0}
print(dd)

{'two': 2}


# 字典的相关函数

In [36]:
# 通用函数：len,max,min,dict
# str(字典)：返回字典的字符串格式
d = {"one":1, "two":2, "three":3}
print(str(d))

{'one': 1, 'two': 2, 'three': 3}


In [37]:
# items:返回字典的键值对组成的元祖格式

d = {"one":1, "two":2, "three":3}
i = d.items()
print(type(i))
print(i)
# 生成一个可迭代的类型

<class 'dict_items'>
dict_items([('one', 1), ('two', 2), ('three', 3)])


In [38]:
# keys: 返回字典的键组成的一个结构
k = d.keys()
print(type(k))
print(k)
# 同理可证values,一个可迭代结构

<class 'dict_keys'>
dict_keys(['one', 'two', 'three'])


In [43]:
# get: 根据指定键返回相应值，它的好处是可以设置默认值
d = {"one":1, "two":2, "three":3}
print(d.get("one"))
# 默认值式none，可以设置
print(d.get("one333", 100))

print(d.get("one333"))
print(d["one333"])

1
100
None


KeyError: 'one333'

In [44]:
# fromkeys: 使用指定的序列作为键，使用一个值作为字典的所有键的值
l = ['eins','zwei','dree']
# 注意fromkeys俩个参数的类型
d = dict.fromkeys(l,"hahahaha")
print(d)

{'eins': 'hahahaha', 'zwei': 'hahahaha', 'dree': 'hahahaha'}


# str的用法

In [10]:
help(str)

Help on class str in module builtins:

class str(object)
 |  str(object='') -> str
 |  str(bytes_or_buffer[, encoding[, errors]]) -> str
 |  
 |  Create a new string object from the given object. If encoding or
 |  errors is specified, then the object must expose a data buffer
 |  that will be decoded using the given encoding and error handler.
 |  Otherwise, returns the result of object.__str__() (if defined)
 |  or repr(object).
 |  encoding defaults to sys.getdefaultencoding().
 |  errors defaults to 'strict'.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(self, format_spec, /)
 |      Return a formatted version of the string as described by format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  

In [11]:
# capitalize 首字母大写，其余小写，返回字符串
s = "i love wangxiaokjing"
print(s.capitalize())

I love wangxiaokjing


In [12]:
# title()将每个单词的首写字母变为大写，返回的是字符串

s = "i love wangxiaojing"
s1 = s.title()
print(s1)

I Love Wangxiaojing


In [13]:
# upper() /lower()将所有字母变为大/小写字母，返回的是字符串
s = "i love wangxiaojing"
s1 = "I LOVE WANGXIAOJING"
print(s.upper())
print(s1.lower())

I LOVE WANGXIAOJING
i love wangxiaojing


In [14]:
# swapcase()大小写互换
# len()计算字符串长度，不属于字符串的内建函数
# len统计长度是按字符个数统计，一个汉子的长度为1
s = "阿大师大师大师大师答"
s1 = "I LIKE 狗生"
print(len(s))
print(len(s1))

10
9


In [17]:
# find()查找指定字符串，找不到返回-1 第一次找到返回第一次索引值
# index()查找指定字符串，找不到报错
s = "asadasdadasdsaffaff"
s1 = s.find('s',3)
s2 = s.index('a')
print(s1)
print(s2)
help(str.find)
help(str.index)

5
0
Help on method_descriptor:

find(...)
    S.find(sub[, start[, end]]) -> int
    
    Return the lowest index in S where substring sub is found,
    such that sub is contained within S[start:end].  Optional
    arguments start and end are interpreted as in slice notation.
    
    Return -1 on failure.

Help on method_descriptor:

index(...)
    S.index(sub[, start[, end]]) -> int
    
    Return the lowest index in S where substring sub is found, 
    such that sub is contained within S[start:end].  Optional
    arguments start and end are interpreted as in slice notation.
    
    Raises ValueError when the substring is not found.



In [18]:
# count() 计算字符串出现次数，返回整数
s = "asdasdfasdfsdafadsfa"
print(s.count('s'))

5


In [21]:
# startswith()检查是否以指定字母开头，分会布尔值
# endswith()检查是否以指定字母结束
s = 'I LIKE DOG'
print(s.startswith("i"))
print(s.endswith("G"))

False
True


In [22]:
# isupper()检测说有字母是否都是大写字母，返回布尔值
# islower
# istitle
#isspace()检测字符串是否是空字符串
s = ' ' 
s1 = "i  asd"
s2 = " "# 至少有一个，否则返回false
print(s.isspace())
print(s1.isspace())
print(s2.isspace())

True
False
True


In [26]:
# isalpha()检测字符串时候是字母组成 返回布尔值
# 说明：汉子在英文字符包裹中被单座字符处理
s = 'i 够like dog'
s1 = 'i够likedog'
print(s.isalpha())
print(s1.isalpha())
# isalnum()检测字符串是否由字母加数字组成或全是字母或全是数字，返回布尔值
# 下面三个检测是否全是由数字组成，区别看例子
# isdigit()
# isdecimal()
# isnumeric()
s = '123'
print(s.isdigit())
print(s.isdecimal())
print(s.isnumeric())
print('='*20)
s = b'101100'
print(s.isdigit())
# print(s.isdecimal())
# print(s.isnumeric())
print('='*20)
s = '123.2'
print(s.isdigit())
print(s.isdecimal())
print(s.isnumeric())
print('='*20)
s = '叁壹佰'
print(s.isdigit())
print(s.isdecimal())
print(s.isnumeric())
print('='*20)
s = "III"# 罗马数字三
print(s.isdigit())
print(s.isdecimal())
print(s.isnumeric())
print('='*20)

False
True
True
True
True
True
False
False
False
False
False
True
False
False
False


In [31]:
help(str.isdigit)
help(str.isnumeric)
help(str.isdecimal)

Help on method_descriptor:

isdigit(self, /)
    Return True if the string is a digit string, False otherwise.
    
    A string is a digit string if all characters in the string are digits and there
    is at least one character in the string.

Help on method_descriptor:

isnumeric(self, /)
    Return True if the string is a numeric string, False otherwise.
    
    A string is numeric if all characters in the string are numeric and there is at
    least one character in the string.

Help on method_descriptor:

isdecimal(self, /)
    Return True if the string is a decimal string, False otherwise.
    
    A string is a decimal string if all characters in the string are decimal and
    there is at least one character in the string.



In [32]:
# split()用指定字符切割字符串 返回由字符串组成的列表
s = '日照香炉生紫烟*疑是银河落九天'
list1 = s.split('*')
print(list1)

['日照香炉生紫烟', '疑是银河落九天']


In [33]:
# splitlines()  以换行切割字符串
s = '''日照香炉生紫烟\n疑似银\n河落九天
飞流直下三千尺'''
print(s.splitlines())

['日照香炉生紫烟', '疑似银', '河落九天', '飞流直下三千尺']


In [34]:
# join()将列表按照指定字符串连接  返回的是字符串
list1 = ['日照香炉生紫烟','疑是银河落九天','飞流直下三千尺']
s = '*'.join(list1)
print(s)

日照香炉生紫烟*疑是银河落九天*飞流直下三千尺


In [35]:
# ljust() 指定字符串长度，内容靠左边，不足的位置用指定字符串填充，默认空格，返回字符串
s = 'abc'
print(len(s))
print(s.ljust(5)+'a')

3
abc  a


In [40]:
s = 'abc'
# center() 指定字符串长度，内容居中，不足的位置用指定字符串填充，默认空格，返回字符串
print(s.center(5,'#'))

#abc#


In [41]:
# rjust() 指定字符串长度，内容靠右，不足的位置用指定字符串填充，默认空格，返回字符串
print(s.rjust(5,'#'))

##abc


In [43]:
# strip()去掉左右俩边指定字符，默认是去掉空格
# lstrip()去掉左侧指定字符，默认空格
# rstrip()去掉右侧指定字符，默认空格
s = '    abc '
print('---'+s.strip()+'--')
print('--'+s+'--')
s = 'aaabcc'
print(s.lstrip('a'))
print(s.rstrip('c'))

---abc--
--    abc --
bcc
aaab


In [44]:
# zfill()指定字符串长度，内容靠右，不足的位置用0填充
s = 'abc'
print(s.zfill(5))

00abc


In [46]:
# maketrans()生成用于字符串替换的映射表
# translate()进行字符串替换
s = '今晚我吃的是小炒肉，可好吃了'
table = s.maketrans('小炒肉','大白菜')
print(table)
print(s.translate(table))

{23567: 22823, 28818: 30333, 32905: 33756}
今晚我吃的是大白菜，可好吃了


In [47]:
s = '今晚我吃的是小炒肉，可好吃了'
# 长度要一样，否则报错
table = s.maketrans('小炒肉','大白菜和粉条')
print(table)
print(s.translate(table))

ValueError: the first two maketrans arguments must have equal length

In [1]:
help(list)

Help on class list in module builtins:

class list(object)
 |  list(iterable=(), /)
 |  
 |  Built-in mutable sequence.
 |  
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self))

In [2]:
# append()向列表末尾添加新援色，返回None
list1 = [1,2,3,4,5,6]
print(id(list1))
list1.append(7)
print(id(list1))
print(list1)

4355360200
4355360200
[1, 2, 3, 4, 5, 6, 7]


In [3]:
# copy()赋值列表
list1 = [1,2,3,4,5,6]
list2 = list1.copy()
print(id(list1))
print(id(list2))

4355360072
4355359560


In [6]:
# count()计算某个元素在列表中出现的次数
list1 = [1,1,2,3,5,4,1]
print(list1.count(1))

3


In [14]:
# extend()讲一个列表继承另一个列表
list1 = [1,2,3,4,5]
print(id(list1))
list2 = [6,7,8,9,10]
print(list2)
list3 = list1.extend(list2)
print(id(list1))
print(list1)
print(list2)
print(list3)
print(list1 + list2)

4355398216
[6, 7, 8, 9, 10]
4355398216
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
4355396552
[6, 7, 8, 9, 10]
None
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 6, 7, 8, 9, 10]


In [12]:
help(list.extend)

Help on method_descriptor:

extend(self, iterable, /)
    Extend list by appending elements from the iterable.



In [15]:
# index()  获取值在列表中的索引
list1 = [1,1,2,3,4,5,6]
print(list1.index(3))

3


In [16]:
print(list1.index(3,2,5))

3


In [17]:
help(list.index)

Help on method_descriptor:

index(self, value, start=0, stop=9223372036854775807, /)
    Return first index of value.
    
    Raises ValueError if the value is not present.



In [25]:
# insert() 在指定位置前插入元素 ，2个参数
list1 = [1,2,3,4,5,6]
print(list1.insert(3,3))
help(list.insert)

None
Help on method_descriptor:

insert(self, index, object, /)
    Insert object before index.



In [33]:
# pop 根据索引译出列表内的一个元素，不给索引默认移除最后一个，返回移除哪个值
list1 = [1,2,3,4,5,6,7,8]
list1.pop(2)
print(list1)
print(list1.pop())
list1.pop()
print(list1)
print(list1.pop())
list1.pop()
print(list1)
print(list1.pop())
list1.pop()
print(list1)
print(list1.pop())
list1.pop()
print(list1)
print(list1.pop())


[1, 2, 4, 5, 6, 7, 8]
8
[1, 2, 4, 5, 6]
6
[1, 2, 4]
4
[1]
1


IndexError: pop from empty list

In [34]:
# remove()  移除列表中指定值
list1 = ['a','b','c','d']
print(list1.remove('b'))
print(list1)

None
['a', 'c', 'd']


In [35]:
# reverse()列表翻转
list1 = [1,2,3,4]
print(id(list1))
list1.reverse()
print(id(list1))
print(list1)

4356032136
4356032136
[4, 3, 2, 1]


In [37]:
# sort() 排序，默认从小到大
list1 = [5,2,4,8,9,4,7]
list1.sort()
print(list1)
# 倒着拍着
list1.sort(reverse=True)
print(list1)

[2, 4, 4, 5, 7, 8, 9]
[9, 8, 7, 5, 4, 4, 2]


In [38]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Stable sort *IN PLACE*.



In [39]:
help(tuple)

Help on class tuple in module builtins:

class tuple(object)
 |  tuple(iterable=(), /)
 |  
 |  Built-in immutable sequence.
 |  
 |  If no argument is given, the constructor returns an empty tuple.
 |  If iterable is specified the tuple is initialized from iterable's items.
 |  
 |  If the argument is a tuple, the return value is the same object.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __getnewargs__(self, /)
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |      Return hash(self).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __le__(self, value, /)
 |

In [40]:
help(set)

Help on class set in module builtins:

class set(object)
 |  set() -> new empty set object
 |  set(iterable) -> new set object
 |  
 |  Build an unordered collection of unique elements.
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __contains__(...)
 |      x.__contains__(y) <==> y in x.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iand__(self, value, /)
 |      Return self&=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  __ior__(self, value, /)
 |      Return self|=value.
 |  
 |  __isub__(self, value, /)
 |      Return self-=value.
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __ixor__(self, value, /)
 |      Re

In [41]:
help(dict)

Help on class dict in module builtins:

class dict(object)
 |  dict() -> new empty dictionary
 |  dict(mapping) -> new dictionary initialized from a mapping object's
 |      (key, value) pairs
 |  dict(iterable) -> new dictionary initialized as if via:
 |      d = {}
 |      for k, v in iterable:
 |          d[k] = v
 |  dict(**kwargs) -> new dictionary initialized with the name=value pairs
 |      in the keyword argument list.  For example:  dict(one=1, two=2)
 |  
 |  Methods defined here:
 |  
 |  __contains__(self, key, /)
 |      True if the dictionary has the specified key, else False.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __init__(self,

In [42]:
# clear()清除整个字典，返回None
# copy()复制字典
# fromkeys()按照指定的序列为键创建字典
list1 = ['a','b','c']
dict1 = {}.fromkeys(list1)
dict2 = {}.fromkeys(list1, 3)
print(dict1,dict2)

{'a': None, 'b': None, 'c': None} {'a': 3, 'b': 3, 'c': 3}


In [47]:
# get()根据键获取指定的值， 找不到的键默认值返回默认值，没有默认值，返回None
dict1 = {'a': 1, 'b': 2, 'c': 3}
print(dict1.get('a'))
print(dict1.get('f'))
print(dict1.get('d',4))

1
None
4


In [53]:
# items()将字典变成元祖的形式方便遍历
dict1 = {'a': 1, 'b': 2, 'c': 3}
for k,v in dict1.items():
    print(k,v)

for i in dict1.items():
    print(i)
    
print(dict1.items())

a 1
b 2
c 3
('a', 1)
('b', 2)
('c', 3)
dict_items([('a', 1), ('b', 2), ('c', 3)])


In [59]:
# pop()移除字典中的指定元素 返回键所对应的值，如果键不存在，返回默认值，没有默认值，则会报错
# popitem()移除键值对，返回键值
# setdefault()在字典里去一个值，如果没有放进去
# update()修改字典中的值
dict1 = {'a': 1, 'b': 2, 'c': 3}
dict1 = dict1.update({'a': 2,'e': 6})
print(dict1)

None


In [62]:
list1 = [1,2,2,4,5,'a']
a = set(list1)
print(a)

{1, 2, 4, 5, 'a'}


In [63]:
# add()向集合中添加元素
set1 = {1,2,3,4,5}
set1.add(6)
print(set1)

{1, 2, 3, 4, 5, 6}


In [64]:
# clear()清空合集
# copy（）复制合集
# pop（）随机弹出一个元素
a = {'a','b','f',4}
a.pop()

'a'

In [ ]:
# remove()删除集合中的某个值，如果这个值不在集合中会报错
# discard()删除某个值，没有啥也不干
# difference()差集
# differnce_update() 区别就是第一个返回一个新的集合，第二个是吧原来的集合覆盖

# 编码问题
- 为什么需要编码
    - 本质上计算机智能识别01代码
    - 如何用一长串代码表示复杂信息 每一个字母->数字
    - 编码-encode  解码-decode
- 编码简史
    - 二进制
        - bit: 一个0或者1的二进制数字
        - byte: 八个01是一个字节
        - 0-127位->英文或者字符
    - 第一阶段：ASCII（英语国家）
    - 第二阶段：百花齐放，gb2,gbk,big5(中国）latin1(欧洲）jis(日本）
        - 乱码就是解码系统和编码系统不同
        - 由于0-127就是给英语的，所以任何编码系统都能解析英语编码
        - ansi-MBCS(multi-bytes charecter set,多子节字符集）（微软）
    - 第三阶段：unicode(ISO)全球统一标准
- 编码的表示方法
- ascii-american standard code for in formation interchange
    - 所有的控制字符（包括回车，删除等）编码再0-31范围内和127
    - 所有标点符号，英文大小写放在32-126之间
    - 预留128-255之间预留
    - 0xxx xxxx是它的编码形式
    
- latin1
    - 0-127不懂，那么可以兼容ascii
    - 128-255位置全部用完，二进制位1xxx xxxx
        - 128-159之间为控制字符
        - 160-255位文字符号
        - 其中包括西欧，希腊语。泰语，阿拉伯语，希伯来语
    - 欧元符号
- gbxxxxxxxxx
    - gb2312
        - 如果一个字节中第一位为0，那么这就是一个ascii
        - 如果第一个字节中第一位为1，那么这是个汉子
        - 1xxxxxxxxx xxxxxxxxx：表示问汉子
    - gbk
    - gb18030
        -2/4位混编

# Unicode
    - 只是一个码表，具体实现没有规定
    - 0-0x10ffff来映射这些这些字符，最多可以容纳1114112个字符
    - 中文的编码范围4e00-4fcf,其中9fc4-9fcf之间的区间没有使用
    - 上述区间全部都是汉子，不包含全角字符，不包含特殊字符
    - utf = unicodetransformationformat
    - utf-8
    
    
                    0x0000~0x007f(0~127)    1字节   0xxxxxxx
                    0x0080~0x07ff(128~2047)    110xxxxx 10xxxxxx
                    0x0800～ffff(2048~65535   三字节  1110xxxx 10xxxxxxx 10xxxxxx
                    0x10000~1ffffff(65536~2097152)  4字节   11110xxx 10xxxxxx 10xxxxxx 10xxxxxxx
                    0x2000000~0x3ffffff   5字节
                    0x4000000～0x7fffffff 
                    
                    
                    

# 习题课（循环，函数）

In [30]:
# 打印出一个矩形
# 控制行
for i in range(1,5):
    # 控制列
    for j in range(1,6):
        print(' * ',end='')
    # 当第i行的5列输出完后进行换行
    print()
print('='*20)
# 打印一个空心的矩形
for i in range(1,5):
    for j in range(1,6):
        if i == 1 or i == 4:
            print(' * ',end='')
        if i == 2 or i == 3:
            if j == 1 or j == 5:
                print(' * ',end='')
            else:
                print('   ',end='')
    print()

 *  *  *  *  * 
 *  *  *  *  * 
 *  *  *  *  * 
 *  *  *  *  * 
 *  *  *  *  * 
 *           * 
 *           * 
 *  *  *  *  * 


In [43]:
# 迭代
# 类似于站的先进后出模式
# 1.要有递推关系
# 2.要有临界值
def digui(num):
    print(' $ '+str(num))
    # 临界值
    if num > 0:
        # 这里用的是调用本身的函数
        digui(num-1)
    else:
        print('='*20)
    # 在判断区间外的
    print(num)
digui(20)
# 传入20， 打印20，进入判断，后面程序进入等待不执行，，print（2）

 $ 20
 $ 19
 $ 18
 $ 17
 $ 16
 $ 15
 $ 14
 $ 13
 $ 12
 $ 11
 $ 10
 $ 9
 $ 8
 $ 7
 $ 6
 $ 5
 $ 4
 $ 3
 $ 2
 $ 1
 $ 0
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
